Add avatar generation service and database support

SAM19991 opened the pull request:
Черновой вариант без тестов

SAM19991 committed:
Add avatar generation service and database support

Introduce `AvatarService` and its implementation, enabling avatar URL generation using DiceBear integration. Add new `avatar_url` column to the `users` table and update configurations to include DiceBear API details.

Hitman-nn commented on src/main/java/school/faang/user_service/config/DiceBearConfig.java:
```diff
@@ -0,0 +1,20 @@
+
+    @NotBlank
+    private String style;
+}
```
Я бы наверное сделал так: без валидации, но со значением по умолчанию + с записью в лог, что подгрузили дефолтные значения. 
Почему: функционал не критичный, старт даже без конфига

Hitman-nn replied:
может стоить расширить конфиг и добавить формат/тип файла авы

Hitman-nn commented on src/main/java/school/faang/user_service/entity/User.java:
```diff
@@ -65,6 +65,9 @@ public class User {
 
+    @Column(name = "avatar_url")
+    private String avatarUrl;
+
```
Почему не UserProfilePic? по названию очень подходит под это дело)

SAM19991 replied:
убрал 
@Column(name = "avatar_url")
private String avatarUrl;
буду использовать userProfilePic

Hitman-nn commented on src/main/java/school/faang/user_service/service/avatar/AvatarServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+
+@Service
+@RequiredArgsConstructor
+public class AvatarServiceImpl implements AvatarService {
```
Может поменять название на более конкретное, чтоб было понятно какой сервис представляет, например как-нибудь DiceBearAvatarServiceImpl

Hitman-nn commented on src/main/java/school/faang/user_service/service/avatar/AvatarServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+            seedSource = String.valueOf(user.getId());
+        } else {
+            seedSource = user.getEmail() != null ? user.getEmail() : "defaultUserSeed";
+        }
```
Мне кажется тут достаточно username (username varchar(64) UNIQUE NOT NULL) и дефолтного значения. Зачем так широко проверять.

Hitman-nn replied:
всякие спец символы, пробелы не сломают ничего? проверял?

Hitman-nn commented on src/main/resources/db/changelog/changeset/user_V009__add_avatar_url.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE users
+    ADD COLUMN avatar_url VARCHAR(1024) NULL;
```
При миграции не стоит выставить дефолтную аву уже созданным юзерам?

SAM19991 committed:
- Renamed the avatar service implementation to DiceBearAvatarServiceImpl for clarity.
- Updated the service to generate the avatar URL using only the username (or a default seed) with URL encoding.
- Removed usage of the standalone avatarUrl field and now store the generated URL into the embedded UserProfilePic.smallFileId field.
- Updated configuration with default values and added log warnings when defaults are used."
- Created a test

SAM19991 committed:
исправил некоторые ошибки

В DiceBearAvatarServiceImpl теперь при отсутствии указанного стиля используется "pixel-art".
Тесты DiceBearAvatarServiceImplTest обновлены для проверки корректного формирования URL с использованием этого значения.

Hitman-nn commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarService.java:
```diff
@@ -0,0 +1,7 @@
+
+import school.faang.user_service.entity.User;
+
+public interface DiceBearAvatarService {
```
Зачем тут поменял название?

Hitman-nn commented on src/main/resources/application.yaml:
```diff
@@ -34,4 +34,11 @@ project-service:
\ No newline at end of file
+  port: 9080
+
+dicebear:
+  api-url: https://api.dicebear.com/9.x
+  style: pixel-art
+
+default-avatar:
+  url: /images/default-avatar.svg
```
а почему не в одном блоке?

Hitman-nn commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+        String apiUrl = diceBearConfig.getApiUrl();
+        String style = diceBearConfig.getStyle();
+
+        if (style == null || style.isEmpty()) {
+            style = "pixel-art";
+            log.warn("DiceBearConfig.style is missing; using default: {}", style);
+        }
+        if (apiUrl == null || apiUrl.isEmpty()) {
+            apiUrl = "https://api.dicebear.com/9.x";
+            log.warn("DiceBearConfig.apiUrl is missing; using default: {}", apiUrl);
+        }
```
1) зачем проверку на дефолт и логирование делать каждый раз при генерации авы, стоит вынести на постконстракт конфига
2) Так ты же дефолтные значения выставляешь в конфиге, тут то зачем?

Hitman-nn commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+
+        String avatarUrl =  String.format("%s/%s/svg?seed=%s", apiUrl, style, encodedSeed);
+
+        if (user.getUserProfilePic()==null)  {
+            user.setUserProfilePic(new UserProfilePic());
+        }
+        user.getUserProfilePic().setSmallFileId(avatarUrl);
```
это не дело данного сервиса. нарушаем solid

Hitman-nn commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+public class DiceBearAvatarServiceImpl implements DiceBearAvatarService {
+
+    private final DiceBearConfig diceBearConfig;
+    private static final String DEFAULT_SEED = "defaultUserSeed";
```
ты же это в конфиг вынес. Либо там, либо тут)

Hitman-nn reviewed (commented):
посмотрел, что-то все перемешал...

SAM19991 committed:
- Переименован `DiceBearAvatarService` в для большей общности. `AvatarService`
- Упрощена реализация : удалена логика обработки значений по умолчанию для URL и стиля API (теперь это ответственность ) и убрано изменение состояния объекта `User` (установка ). `DiceBearAvatarServiceImpl``DiceBearConfig``UserProfilePic`
- Обновлены тесты для соответствия изменениямю.`

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public String generateAvatarUrl(User user) {
+        String seedSource;
+        if (user == null || user.getUsername() == null || user.getUsername().isEmpty()) {
+            seedSource = DEFAULT_SEED;
+        } else {
+            seedSource = user.getUsername();
+        }
+        String encodedSeed = URLEncoder.encode(seedSource, StandardCharsets.UTF_8);
+        String apiUrl = diceBearConfig.getApiUrl();
+        String style = diceBearConfig.getStyle();
+
+        return String.format("%s/%s/svg?seed=%s", apiUrl, style, encodedSeed);
```
мне кажется ты не совсем понял суть задачи.
нам требуется выполнить запрос в сторонний сервис для получения аватара.
то-есть тебе потребуется сделать feignClient, который будет выполнять запрос для получения аватара, а полученный аватар сохранять у нас.
В описании еще есть "Если же сервис предоставляет просто URL к случайному аватару, то достаточно будет просто сохранить его в нашу БД для пользователя." возможно у https://www.dicebear.com/ есть api, которое возвращает не аватар а ссылку на него, тогда мы должны сделать запрос в dicebear, получить от него ссылку и сохранить

SAM19991 committed:
- API для создания и получения аватаров
- Интеграция с DiceBear для генерации картинок
- Сохранение ссылок на аватары в базу данных
- Обработка ошибок и валидация
- Оставил только URL-подход для генерации аватаров, так как у нас нет аккаунта на Amazone S3

miked4king commented on src/main/java/school/faang/user_service/controller/avatar/AvatarController.java:
```diff
@@ -0,0 +1,33 @@
+
+    @GetMapping("/{userId}/avatar")
+    public AvatarDto getUserAvatar(@PathVariable @Positive(message = "User ID must be positive") Long userId) {
+        return avatarService.getAvatarDto(userId);
```
в названии метода не обязательно указывать dto)
поправь в методе ниже тоже пж

miked4king commented on src/main/java/school/faang/user_service/dto/avatar/AvatarDto.java:
```diff
@@ -0,0 +1,16 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
не очень нравится использование Builder, он предназначен для конструирования больших или сложных объектов, 3 поля не кажется большим)

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -15,20 +21,145 @@
+    }
+
+    private void validateUserId(Long userId) {
+        if (userId == null || userId <= 0) {
```
а вот в контроллере ты добавил аннотацию Positive, разве до сюда дойдет null или <0 ?
кажется эту проверку можно убрать

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -15,20 +21,145 @@
+
+    private User findUserById(Long userId) {
+        return userRepository.findById(userId)
+                .orElseThrow(() -> new DataValidationException("User not found with id: " + userId));
```
EntityNotFoundException, т.к в базе не находим

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -15,20 +21,145 @@
+    private final AvatarMapper avatarMapper;
 
     @Override
+    public AvatarDto getAvatarDto(Long userId) {
```
getAvatarDto - название не соответствует реальности на мой взгляд, вроде хотим получить аватар, но при этом обновляем что то у юзера

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -15,20 +21,145 @@
+        if (user == null) {
+            log.error("User is null!");
+            throw new IllegalArgumentException("User cannot be null");
+        }
```
а он может быть null? вроде из репозитория его получаешь и если не найдешь бросишь исключение

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -15,20 +21,145 @@
+            log.info("Step 3: Built avatar URL: {}", avatarUrl);
+
+            log.info("Step 4: Saving avatar to user...");
+            saveAvatarToUser(user, avatarUrl);
```
вроде метод называется generateAvatarUrl, а тут мы его еще и сохраняем, давай подумаем над названием метода, либо уберем сохранение в бд

miked4king reviewed (changes requested):
В целом гораздо лучше, давай почистим логи в сервисе, их слишком много.
подумаем над названиями методов или порефакторим немного, потому что получается в некоторых местах getAvatar, а на деле много чего происходит по типу сохранения.

SAM19991 committed:
- Переименовал методы (убрал "Dto" из названий)
- Удалил лишние проверки и валидацию
- Сильно сократил логирование - оставил только важное
- Разделил методы по ответственности
- Поменял исключения на более подходящие
- Обновил контроллер под новые названия

miked4king commented on src/main/java/school/faang/user_service/controller/avatar/AvatarController.java:
```diff
@@ -0,0 +1,33 @@
+import school.faang.user_service.service.avatar.AvatarService;
+
+@RestController
+@RequestMapping("/users")
```
давай будем придерживаться лучших практик и добавим версионирование, получится /api/v1/users

SAM19991 replied:
готово

miked4king commented on src/main/java/school/faang/user_service/service/avatar/AvatarService.java:
```diff
@@ -0,0 +1,14 @@
+import school.faang.user_service.entity.User;
+
+public interface AvatarService {
+    String generateAvatarUrl(User user);
```
данный метод можем убрать из интерфейса, т.к он вызывается только в рамках сервиса DiceBearAvatarServiceImpl

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -0,0 +1,120 @@
+    public AvatarDto generateAvatar(Long userId) {
+        log.info("Generating avatar for user: {}", userId);
+
+        try {
```
не вижу чтобы по коду выбрасывались проверяемые исключения, можем убрать блоки try/catch в этом методе, generateAvatarUrl, saveAvatarToUser

SAM19991 replied:
готово

miked4king commented on src/main/java/school/faang/user_service/service/avatar/DiceBearAvatarServiceImpl.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private String buildAvatarUrl(String seed) {
+        if (seed == null || seed.trim().isEmpty()) {
+            seed = diceBearConfig.getDefaultSeed();
+        }
```
этот блок можем убрать, т.к ты вызываешь данный метод из generateAvatarUrl, и ранее данные проверки уже были

SAM19991 replied:
готово

miked4king reviewed (changes requested):
это последняя итерация ревью, поправим 4 коммента и можем мержить)
только предварительно разреши конфликты, тебе потребуется на свою ветку подтянуть pull вашу мастер ветку и решить конфликты в классе RestControllerExceptionHandler

SAM19991 committed:
refactor: remove unnecessary try-catch blocks and redundant validations

SAM19991 committed:
resolve merge conflicts in RestControllerExceptionHandler

SAM19991 committed:
resolve merge conflicts in RestControllerExceptionHandler
